neutromeratio/parameter_gradients.py

```diff
@@ -3,57 +3,189 @@
+        lambda0_e = [e/kT for e in [e_b_stddev[0] for e_b_stddev in lambda0_e_b_stddev]]
+        lambda1_e = [e/kT for e in [e_b_stddev[0] for e_b_stddev in lambda1_e_b_stddev]]
 
-            new_snapshots = list(traj[equil::thinning].xyz * unit.nanometer)[:max_snapshots_per_window]
-            N_k.append(len(new_snapshots))
-            snapshots.extend(new_snapshots)
+        def get_mix(lambda0, lambda1, lam=0.0):
+            return (1 - lam) * np.array(lambda0) + lam * np.array(lambda1)
 
-        self.snapshots = snapshots
-        N = len(snapshots)
-        u_kn = np.zeros((K, N))
-        for k in range(K):
-            lamb = lambdas[k]
-            for n in range(N):
-                u_kn[k, n] = self.ani_model.calculate_energy(snapshots[n], lamb) / kT
+        print('Nr of atoms: {}'.format(n_atoms))
+        
+        u_kn = np.stack(
+            [get_mix(lambda0_e, lambda1_e, lam) for lam in sorted(used_lambdas)]
+            )
+        
         self.mbar = MBAR(u_kn, N_k)
 
-    def compute_perturbed_free_energies(self, u_ln):
+    def remove_confs_with_high_stddev(self, max_snapshots_per_window:int, per_atom_thresh:float):
+        
+        """
+        Removes conformations with ensemble energy stddev per atom above a given treshold.
+        Parameters
+        ----------
+        max_snapshots_per_window : int
+            maximum number of conformations per lambda window
+        per_atom_thresh : float
+            per atom stddev threshold - by default this is set to 0.5 kJ/mol/atom
+        """
+
+        def calculate_stddev(snapshots):
+            # calculate energy, bias and stddev for endstates
+            lambda0_e_b_stddev = [self.ani_model.calculate_energy(x, lambda_value=0.0) for x in tqdm(snapshots)]
+            lambda1_e_b_stddev = [self.ani_model.calculate_energy(x, lambda_value=1.0) for x in tqdm(snapshots)]
+
+            # extract endpoint stddev and return
+            lambda0_stddev = [stddev/kT for stddev in [e_b_stddev[2] for e_b_stddev in lambda0_e_b_stddev]]
+            lambda1_stddev = [stddev/kT for stddev in [e_b_stddev[2] for e_b_stddev in lambda1_e_b_stddev]]
+            return np.array(lambda0_stddev), np.array(lambda1_stddev)
+
+        def compute_linear_penalty(current_stddev):
+            # calculate the total energy stddev treshold based on the provided per_atom_tresh 
+            # and the number of atoms
+            total_thresh = (per_atom_thresh * self.n_atoms) * unit.kilojoule_per_mole
+            # if stddev for a given conformation < total_tresh => 0.0
+            # if stddev for a given conformation > total_tresh => stddev - total_treshold
+            linear_penalty = np.maximum(0, current_stddev - (total_thresh/kT))
+            return linear_penalty
+
+        def compute_last_valid_ind(linear_penalty):
+            # return the idx of the first entry that is above 0.0
+            if np.cumsum(linear_penalty) < 0.1: # means all can be used
```

Comment:
why is the comparison to 0.1? and shouldn't the return value be `len(linear_penalty) - 1` or so?

Reply:
reason for 0.1 is that I didn't want to compare directly to 0.0 since cumsum over a lot of zeros could accumulate some epsilon -- but np.isclose is the cleaner version of this for sure!

we use the return value for slicing -- so -1 is the same as `len(linear_penalty) - 1`